### PR TITLE
[MOD-3177] Differentiate between an expired output and unfinished input

### PR DESCRIPTION
**Pull request by cathyzbn**
## Describe your changes
[[MOD-3177]](https://linear.app/modal-labs/issue/MOD-3117/differentiate-between-an-expired-output-and-unfinished-input)
Return different Timeout errors for unfinished vs. expired function call outputs.
- When the function call outputs are not available, raise TimeoutError
- When the function call outputs are expired in Redis, raise OutputExpiredError

## Backward/forward compatibility checks

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers

Note on protobuf: protobuf message changes in one place may have impact to
multiple entities (client, server, worker, database). See points above.

### CODE REVIEW

File: modal/functions.py [hunk 1]
```diff
@@ -181,7 +185,7 @@ async def poll_function(self, timeout: Optional[float] = None):
         )
 
         if len(items) == 0:
-            raise TimeoutError()
+            raise TimeoutError("Function call has not finished")
```
Comment (mwaskom) [inline modal/functions.py:188]: Is it right that we're reporting a timeout _on the `get`_ operation, rather than on the underlying function call? We may want to reword this to make that a little more clear.

File: modal/exception.py [hunk 2]
```diff
@@ -86,6 +86,10 @@ class DeserializationError(Error):
     """Raised to provide more context when an error is encountered during deserialization."""
 
 
+class OutputExpiredError(Error):
```
Comment (mwaskom) [inline modal/exception.py:89]: Since this is a specific kind of timeout (and because we previously raised `TimeoutError`), you could make it derive from the more general class maybe.

Comment (cathyzbn) [reply modal/exception.py]: Thanks for pointing that out! just changed.

File: modal/functions.py (hunk 1 above)
Comment (cathyzbn) [reply modal/functions.py]: yeah, should be reporting on the get operation

File: modal_proto/api.proto [hunk 3]
```diff
@@ -1069,6 +1069,7 @@ message FunctionGetOutputsResponse {
   repeated int32 idxs = 3;
   repeated FunctionGetOutputsItem outputs = 4;
   string last_entry_id = 5;
+  bool is_expired = 6;
```
Comment (mwaskom) [inline modal_proto/api.proto:1072]: Does this require server-side changes? Normally we would make such client-server updates in three steps:

- Merge new protos in the client
- Merge server side changes
- Merge client-side feature code

Comment (cathyzbn) [reply modal_proto/api.proto]: got it! the server side changes are [here](https://github.com/modal-labs/modal/pull/13423), and i'll make a separate pr that only changes the protos.

File: modal/functions.py [hunk 4]
```diff
@@ -160,6 +162,8 @@ async def pop_function_call_outputs(
                 # update timeout in retry loop
                 backend_timeout = min(OUTPUTS_TIMEOUT, t0 + timeout - time.time())
                 if backend_timeout < 0:
+                    if len(response.outputs) == 0 and response.num_unfinished_inputs == 0:
+                        raise OutputExpiredError()
```
Comment (gongy) [inline modal/functions.py:166]: suggestion: raising here means we change the behavior of `.remote()`. while this is arguably more consistent, maybe it would be better to keep the surface area smaller to just `.get()`?

**Review by gongy — APPROVED** [review]
looks great!

File: test/conftest.py [hunk 5]
```diff
@@ -721,7 +721,7 @@ async def FunctionGetOutputs(self, stream):
 
             await stream.send_message(api_pb2.FunctionGetOutputsResponse(outputs=[output]))
         else:
-            await stream.send_message(api_pb2.FunctionGetOutputsResponse(outputs=[]))
+            await stream.send_message(api_pb2.FunctionGetOutputsResponse(outputs=[], num_unfinished_inputs=-1))
```
Comment (gongy) [inline test/conftest.py:724]: 🤔
